src/app/storage/FabricTableImpl.ipp

```diff
@@ -318,14 +316,18 @@ struct FabricEntryData : public PersistentData<kFabricMaxBytes>
     CHIP_ERROR SaveEntry(PersistentStorageDelegate * storage, const TableEntry & src_entry)
     {
         CHIP_ERROR err = CHIP_NO_ERROR;
-        TypedTableEntryData entry(endpoint_id, fabric_index, src_entry.mStorageId, src_entry.mStorageData);
         // Look for empty storage space
 
-        err = this->Find(entry.mStorageId, entry.index);
+        EntryIndex index;
+        err = this->Find(src_entry.mStorageId, index);
+
+        Store persistentStore;
+        // C++ doesn't have const constructors; variable is declared const
+        const TypedTableEntryData entry(endpoint_id, fabric_index, const_cast<TableEntry &>(src_entry), index);
```

Comment:
I guess this is safe because `entry` is not going to be mutated and hence will not try to mutate `src_entry` even though it cast away `const`?

Nothing actually enforces that behavior on `TypedTableEntryData`....  It might be safer to have two separate classes:

1. ImmutableTypedTableEntryData, which has only the const bits and has a `const TableEntry &` protected member.
2. TypedTableEntryData, which inherits from ImmutableTypedTableEntryData, takes `TableEntry &` as ctor argument, and then passes it to superclass.  The non-const bits live on here and use const_cast as needed on the member, which it knows is safe because it initialzied it with a non-const thing to start with....

Followup is fine for this.

Reply:
So it is enforced because of `const TypedTableEntryData` - it makes it so that we can only call const methods on the resulting object

Ideally if we're constructing a const object as we are here, we could have a const constructor which treats all fields as const... but unfortunately C++ doesn't offer that.

Reply:
> it makes it so that we can only call const methods on the resulting object

That's true, but because the object stores a non-const ref to `src_entry`, nothing prevents a const method on the object from trying to mutate the entry, right?  Apart from discipline by the programmer.  The compiler will absolutely not prevent it:

```
class Foo {
public:
  Foo(int & arg) : mVal(arg) {}
  void DoStuff() const { mVal++; }
  int Value() const { return mVal; }
private:
  int & mVal;
};

int main(int argc, char * argv[])
{
  int z = 5;
  const Foo foo(z);

  foo.DoStuff();
  fprintf(stdout, "Val: %d\n", foo.Value());
  return 0;
}
```

prints 6.

I think I would prefer to have compiler enforcement here, but as I said that should be a separate PR.

Reply:
OK I just tried it, it does compile - that's kinda wild, how are you allowed to call non-const operations on fields? I swore that const correctness propagated to fields, am I just misremembering? 😬

How can a method be const if I can call a mutating method on a field 😅🤣🤦‍♂️

Reply:
Resolving & addressing in follow-up, apologies for thinking this offered enforcement